### PR TITLE
Multi-threaded quantization

### PR DESCRIPTION
**Pull request by ikawrakow**
This PR adds multi-threading for quantization.

The gain is very minor for small models (e.g., LLaMA 7B) and simple quantization (`Q4_0` and `Q4_1`), but very significant for large models and the now more elaborate `Q4_2` quantization.

`quantize-stats` now finishes in just 14.5 seconds (7B) or 44 seconds (13B) on my computer for all 3 quantization types. The single-threaded version took 144 seconds (7B) or 242 seconds (13B).

### CODE REVIEW

**Comment by DannyDaemonic** [comment]
This could make more accurate but slow quantization methods more practical. (See #835.)

File: llama.cpp [hunk 1]
```diff
@@ -1578,6 +1581,8 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
         default: throw format("invalid output file type %d\n", ftype);
     };
 
+    if (nthread <= 0) nthread = std::thread::hardware_concurrency();
+
```
Comment (ggerganov) [inline llama.cpp:1585]:
```suggestion
    if (nthread <= 0) {
        nthread = std::thread::hardware_concurrency();
    }
    
    fprintf(stderr, "%s : using %d threads\n", nthread);
```

File: llama.cpp [hunk 2]
```diff
@@ -1639,21 +1647,37 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
...
-                    } break;
-                case GGML_TYPE_Q4_2:
-                    {
-                        new_size = ggml_quantize_q4_2(f32_data, new_data, nelements, (int) tensor.ne.at(0), hist_cur.data());
-                    } break;
-                default:
-                    LLAMA_ASSERT(false);
+            int chunk_size = 32 * 512;
+            int nchunk = (nelements + chunk_size - 1)/chunk_size;
+            int nthread_use = nthread > 1 ? std::max(1, std::min(nthread, nchunk)) : 1;
```
Comment (ggerganov) [inline llama.cpp:1652]:
```suggestion
            const int chunk_size = 32 * 512;
            const int nchunk = (nelements + chunk_size - 1)/chunk_size;
            const int nthread_use = nthread > 1 ? std::max(1, std::min(nthread, nchunk)) : 1;
```

File: llama.cpp [hunk 3]
```diff
@@ -1639,21 +1647,37 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
...
+                        lock.unlock();
+                        size_t last = std::min(nelements, first + chunk_size);
+                        if (local_hist.empty()) local_hist.resize(hist_cur.size(), 0);
+                        local_size += ggml_quantize_chunk(new_type, f32_data, new_data, first, last - first, local_hist.data());
+                    }
+                };
+                if (int(workers.size()) < nthread_use-1) workers.resize(nthread_use-1);
+                for (int it=0; it<nthread_use-1; ++it) workers[it] = std::thread(compute);
+                compute();
+                for (int it=0; it<nthread_use-1; ++it) workers[it].join();
```
Comment (ggerganov) [inline llama.cpp:1680]:
```suggestion
                if (int(workers.size()) < nthread_use - 1) workers.resize(nthread_use - 1);
                for (int it = 0; it < nthread_use - 1; ++it) workers[it] = std::thread(compute);
                compute();
                for (int it = 0; it < nthread_use - 1; ++it) workers[it].join();
```

File: ggml.c [hunk 4]
```diff
@@ -11870,6 +11870,33 @@ size_t ggml_quantize_q4_2(const float * src, void * dst, int n, int k, int64_t *
...
+                result = ggml_quantize_q4_1(src + start, block, n, n, hist);
+            } break;
+        case GGML_TYPE_Q4_2:
+            {
+                assert (start % QK4_2 == 0);
+                block_q4_2 * block = (block_q4_2*)dst + start / QK4_2;
+                result = ggml_quantize_q4_2(src + start, block, n, n, hist);
+            } break;
+        default:
+            assert(false);
```
Comment (ggerganov) [inline ggml.c:11895]: Change these `assert` to `GGML_ASSERT` so it crashes even in Release if the user provides wrong data

**Review by ggerganov — APPROVED** [review]

**Comment by prusnak** [comment]
Please resolve conflicts with the master branch

**Pull request merged**